internal/worker/charmrevision/worker.go

```diff
@@ -66,63 +138,528 @@ func (config Config) Validate() error {
+	} else if len(metadata) == 0 {
+		return nil
+	}
+
+	// Override the context which will use a shorter timeout for sending
+	// metrics.
+	ctx, cancel := context.WithTimeout(ctx, charmhub.RefreshTimeout)
+	defer cancel()
+
+	if err := client.RefreshWithMetricsOnly(ctx, metadata); err != nil {
+		return internalerrors.Errorf("%w: %w", ErrFailedToSendMetrics, err)
+	}
+
+	return nil
+}
+
+func (w *revisionUpdateWorker) fetchInfo(ctx context.Context, client CharmhubClient, buildTelemetry bool, ids []charmhubID, apps []appInfo) ([]latestCharmInfo, error) {
+	metrics, err := w.buildMetricsMetadata(ctx, buildTelemetry)
+	if err != nil {
+		return nil, internalerrors.Capture(err)
+	}
+
+	// Override the context which will use a shorter timeout for sending
+	// metrics.
+	ctx, cancel := context.WithTimeout(ctx, charmhub.RefreshTimeout)
+	defer cancel()
+
+	response, err := w.request(ctx, client, metrics, ids)
+	if err != nil {
+		return nil, internalerrors.Errorf("requesting latest information: %w", err)
+	}
+
+	if len(response) != len(apps) {
+		return nil, internalerrors.Errorf("expected %d responses, got %d", len(apps), len(response))
+	}
+
+	var latest []latestCharmInfo
+	for i, result := range response {
+		latest = append(latest, latestCharmInfo{
+			url:       apps[i].charmURL,
+			timestamp: result.timestamp,
+			revision:  result.revision,
+			resources: result.resources,
+			appID:     apps[i].id,
+		})
+	}
+
+	return latest, nil
+}
+
+// charmhubLatestCharmInfo fetches the latest information about the given
+// charms from charmhub's "charm_refresh" API.
+func (w *revisionUpdateWorker) request(ctx context.Context, client CharmhubClient, metrics charmhub.Metrics, ids []charmhubID) ([]charmhubResult, error) {
+	configs := make([]charmhub.RefreshConfig, len(ids))
+	for i, id := range ids {
+		base := charmhub.RefreshBase{
+			Architecture: id.arch,
+			Name:         id.osType,
+			Channel:      id.osChannel,
+		}
+		cfg, err := charmhub.RefreshOne(id.instanceKey, id.id, id.revision, id.channel, base)
+		if err != nil {
+			return nil, internalerrors.Capture(err)
+		}
+		cfg, err = charmhub.AddConfigMetrics(cfg, id.metrics)
+		if err != nil {
+			return nil, internalerrors.Capture(err)
+		}
+		configs[i] = cfg
+	}
+	config := charmhub.RefreshMany(configs...)
+
+	ctx, cancel := context.WithTimeout(ctx, charmhub.RefreshTimeout)
+	defer cancel()
+
+	responses, err := client.RefreshWithRequestMetrics(ctx, config, metrics)
+	if err != nil {
+		return nil, internalerrors.Capture(err)
+	}
+
+	results := make([]charmhubResult, len(responses))
+	for i, response := range responses {
+		if results[i], err = w.refreshResponseToCharmhubResult(response); err != nil {
+			return nil, internalerrors.Capture(err)
+		}
+	}
+	return results, nil
+}
+
+func (w *revisionUpdateWorker) getCharmhubClient(ctx context.Context) (CharmhubClient, error) {
+	// Get a new downloader, this ensures that we've got a fresh
+	// connection to the charm store.
```

Comment:
Why a downloader? This just returns a `httpClient`.

Reply:
I did it a different way originally, the comment is obsolete.